File: app/src/main/java/com/mintthursday/recipe/show/ingredients/RecipeIngredientsFragment.java
package com.mintthursday.recipe.show.ingredients;

import android.os.Bundle;
import android.text.Editable;
import android.text.InputFilter;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.mintthursday.DecimalDigitsInputFilter;
import com.mintthursday.R;
import com.mintthursday.models.Ingredient;
import com.mintthursday.models.Recipe;

import java.util.ArrayList;
import java.util.List;


public class RecipeIngredientsFragment extends Fragment {

    private static final String ARG_RECIPE_INGREDIENTS = "ARG_RECIPE_INGREDIENTS";
    private final List<Ingredient> currentSelectedIngredients = new ArrayList<>();
    private double count;
    private Recipe recipe;
    private List<Ingredient> ingredientList;
    private RecyclerView recyclerView;
    private IngredientShowAdapter ingredientShowAdapter;
    private ImageButton btnMinus;
    private ImageButton btnPlus;
    private Button btnAddToCart;
    private EditText countPortions;
    private final TextWatcher countTextWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            String a = s.toString().trim();
            if (!a.isEmpty()) {
                if (a.startsWith(".")) {
                    a = "0" + a;
                    countPortions.setText(a);
                    countPortions.setSelection(countPortions.getText().length());
                }
                double count = Double.parseDouble(a);
                if (count > 0) {
                    List<Ingredient> ingredientList = recipe.getIngredients();
                    List<Ingredient> listNew = new ArrayList<>();
                    for (int i = 0; i < ingredientList.size(); i++) {
                        double qty = recipe.getIngredients().get(i).getQuantity() / recipe.getCountPortion() * count;
                        Ingredient newIngredient = new Ingredient(ingredientList.get(i).getName(), qty, ingredientList.get(i).getUnit());
                        listNew.add(newIngredient);
                        if (!currentSelectedIngredients.isEmpty()) {
                            for (int j = 0; j < currentSelectedIngredients.size(); j++) {
                                if (newIngredient.getName().equals(currentSelectedIngredients.get(j).getName())) {
                                    currentSelectedIngredients.set(j, newIngredient);
                                    Log.i("Mint", currentSelectedIngredients.toString());
                                    break;
                                }
                            }
                        }
                    }
                    ingredientShowAdapter.setItems(listNew);
                    RecipeIngredientsFragment.this.count = count;
                }
            }
        }
    };
    private final View.OnClickListener onClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {

            switch (v.getId()) {
                case R.id.minus:
                    if (count > 1) {
                        count--;
                        countPortions.setText(String.valueOf(count));
                    }
                    break;
                case R.id.plus:
                    count++;
                    countPortions.setText(String.valueOf(count));
                    break;
                default:
                    break;
            }
        }
    };

    public static RecipeIngredientsFragment newInstance(Recipe recipe) {
        RecipeIngredientsFragment fragment = new RecipeIngredientsFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_RECIPE_INGREDIENTS, recipe);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_recipe_ingredients, container, false);

        btnMinus = view.findViewById(R.id.minus);
        btnPlus = view.findViewById(R.id.plus);
        btnAddToCart = view.findViewById(R.id.addToCart);
        countPortions = view.findViewById(R.id.count);

        btnMinus.setOnClickListener(onClickListener);
        btnPlus.setOnClickListener(onClickListener);
        countPortions.addTextChangedListener(countTextWatcher);
        countPortions.setFilters(new InputFilter[]{new DecimalDigitsInputFilter(2)});

        recipe = getArguments().getParcelable(ARG_RECIPE_INGREDIENTS);
        ingredientList = recipe.getIngredients();

        recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        ingredientShowAdapter = new IngredientShowAdapter(ingredientList, new IngredientShowAdapter.OnItemCheckListener() {
            @Override
            public void onItemCheck(Ingredient ingredient) {
                currentSelectedIngredients.add(ingredient);
                btnAddToCart.setEnabled(true);
                Log.i("Mint", currentSelectedIngredients.toString());
            }

            @Override
            public void onItemUncheck(Ingredient ingredient) {
                currentSelectedIngredients.remove(ingredient);
                if (currentSelectedIngredients.isEmpty()){
                    btnAddToCart.setEnabled(false);
                }
                Log.i("Mint", currentSelectedIngredients.toString());
            }
        });

        countPortions.setText(String.valueOf(recipe.getCountPortion()));
        count = recipe.getCountPortion();

        recyclerView.setAdapter(ingredientShowAdapter);

        return view;
    }
}

File: app/src/main/java/com/mintthursday/recipe/show/ingredients/RecipeIngredientsFragment.kt
package com.mintthursday.recipe.show.ingredients

import android.os.Bundle
import android.text.Editable
import android.text.InputFilter
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.mintthursday.DecimalDigitsInputFilter
import com.mintthursday.databinding.FragmentRecipeIngredientsBinding
import com.mintthursday.models.Ingredient
import com.mintthursday.models.Recipe
import com.mintthursday.recipe.show.ingredients.IngredientShowAdapter.OnItemCheckListener
import java.util.*

class RecipeIngredientsFragment : Fragment() {

    private var _binding: FragmentRecipeIngredientsBinding? = null
    private val binding get() = _binding!!

    private val recipe: Recipe by lazy { arguments?.getParcelable(ARG_RECIPE_INGREDIENTS)!! }

    private val currentSelectedIngredients = mutableListOf<Ingredient>()
    private var count = 0.0
    private lateinit var ingredientShowAdapter: IngredientShowAdapter

    private val countTextWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
        override fun afterTextChanged(s: Editable) {
            var a = s.toString().trim { it <= ' ' }
            if (a.isNotEmpty()) {
                if (a.startsWith(".")) {
                    a = "0$a"
                    binding.count.setText(a)
                    binding.count.setSelection(binding.count.text.length)
                }
                val count = a.toDouble()
                if (count > 0) {
                    val ingredientList = recipe.ingredients
                    val listNew = mutableListOf<Ingredient>()
                    for (i in ingredientList.indices) {
                        val qty = recipe.ingredients[i].quantity / recipe.countPortion * count
                        val newIngredient = Ingredient(ingredientList[i].name, qty, ingredientList[i].unit)
                        listNew.add(newIngredient)
                        if (currentSelectedIngredients.isNotEmpty()) {
                            for (j in currentSelectedIngredients.indices) {
                                if (newIngredient.name == currentSelectedIngredients[j].name) {
                                    currentSelectedIngredients[j] = newIngredient
                                    Log.i("Mint", currentSelectedIngredients.toString())
                                    break
                                }
                            }
                        }
                    }
                    ingredientShowAdapter.setItems(listNew)
                    this@RecipeIngredientsFragment.count = count
                }
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {

        _binding = FragmentRecipeIngredientsBinding.inflate(inflater, container, false)

        binding.btnMinus.setOnClickListener({
            if (count > 1) {
                count--
                binding.count.setText(count.toString())
            }
        })
        binding.btnPlus.setOnClickListener({
            count++
            binding.count.setText(count.toString())
        })

        binding.count.addTextChangedListener(countTextWatcher)
        binding.count.setFilters(arrayOf<InputFilter>(DecimalDigitsInputFilter(2)))

        binding.recyclerView.layoutManager = LinearLayoutManager(activity)
        ingredientShowAdapter = IngredientShowAdapter(recipe.ingredients, object : OnItemCheckListener {
            override fun onItemCheck(ingredient: Ingredient) {
                currentSelectedIngredients.add(ingredient)
                binding.btnAddToCart.setEnabled(true)
                Log.i("Mint", currentSelectedIngredients.toString())
            }

            override fun onItemUncheck(ingredient: Ingredient) {
                currentSelectedIngredients.remove(ingredient)
                if (currentSelectedIngredients.isEmpty()) {
                    binding.btnAddToCart.setEnabled(false)
                }
                Log.i("Mint", currentSelectedIngredients.toString())
            }
        })
        binding.count.setText(recipe.countPortion.toString())
        count = recipe.countPortion.toDouble()
        binding.recyclerView.adapter = ingredientShowAdapter
        return binding.root
    }

    companion object {
        private const val ARG_RECIPE_INGREDIENTS = "ARG_RECIPE_INGREDIENTS"
        fun newInstance(recipe: Recipe): RecipeIngredientsFragment {
            val fragment = RecipeIngredientsFragment()
            val bundle = Bundle()
            bundle.putParcelable(ARG_RECIPE_INGREDIENTS, recipe)
            fragment.arguments = bundle
            return fragment
        }
    }
}